engine-src/Game/LambdaHack/Client/UI.hs

```diff
@@ -148,25 +149,30 @@ humanCommand = do
               modifySession $ \sess -> sess {sreportNull = True}
               -- Display base frame at the end.
               return []
-        LastRecord seqCurrent seqPrevious k <- getsSession slastRecord
-        let slastRecord
-              | k == 0 = LastRecord [] seqCurrent 0
-              | otherwise = LastRecord [] (seqCurrent ++ seqPrevious) (k - 1)
-        modifySession $ \sess -> sess {slastRecord}
         leader <- getLeaderUI
         b <- getsState $ getActorBody leader
         when (bhp b <= 0 && Just leader /= mOldLeader) $ displayMore ColorBW
           "If you move, the exertion will kill you. Consider asking for first aid instead."
         km <- promptGetKey ColorFull (EM.fromList [(propFont, over)]) False []
+        recording <- getsSession srecording
+        when recording $ do
+           -- If we're recording a in-game macro, we want to record each key.
+           macro <- getsSession smacroBuffer
+           modifySession $ \sess -> sess {smacroBuffer = km : macro}
         abortOrCmd <- do
           -- Look up the key.
           CCUI{coinput=InputContent{bcmdMap}} <- getsSession sccui
           case km `M.lookup` bcmdMap of
+            -- We can repeat every last action except 'repeat last action' action,
+            -- so we don't record that in last action's buffer.
+            Just (_, _, RepeatLast) -> cmdHumanSem RepeatLast
```

Comment:
I suspect the swaitTimes hack is needed even with RepeatLast, but we ignore it here.

Reply:
Yeah, I wasn't quite sure here. Can you remember what it exactly does?

Reply:
As seen in this code, it zeroes swaitTimes after two keystrokes. Elsewhere, if the keystroke invokes a waiting command, the zeroing is interrupted (the size is again made positive) and this counter of consecutive waiting commands is incremented. So, in effect, it zeroes swaitTimes after any non-waiting keystroke (perhaps with a 1 keystroke delay).

Feel free to unhack. :)

Reply:
Whoa, I don't feel adventurous enough today :-D